src/gridfs/upload.rs

```diff
@@ -0,0 +1,154 @@
+        let mut n = 0;
+
+        loop {
+            let mut buf = vec![0u8; chunk_size as usize];
+            let bytes_read = match source.read(&mut buf).await {
+                Ok(0) => break,
+                Ok(n) => n,
+                Err(error) => {
+                    self.chunks()
+                        .delete_many(doc! { "files_id": &files_id }, None)
+                        .await?;
+                    return Err(ErrorKind::Io(error.into()).into());
+                }
+            };
+
+            buf.truncate(bytes_read);
+            let chunk = Chunk {
+                id: ObjectId::new(),
+                files_id: files_id.clone(),
+                n,
+                data: RawBinaryRef {
+                    subtype: BinarySubtype::Generic,
+                    bytes: &buf,
+                },
+            };
+            self.chunks().insert_one(chunk, None).await?;
+
+            length += bytes_read as u64;
+            n += 1;
+        }
+
+        let file = FilesCollectionDocument {
+            id: files_id,
+            length,
+            chunk_size,
+            upload_date: DateTime::now(),
+            filename: Some(filename.as_ref().to_string()),
+            metadata: options.and_then(|opts| opts.metadata),
+        };
+        self.files().insert_one(file, None).await?;
+
+        Ok(())
+    }
+
+    async fn create_indexes(&self) -> Result<()> {
+        if !self.inner.created_indexes.load(Ordering::SeqCst) {
+            let find_options = FindOneOptions::builder()
+                .selection_criteria(SelectionCriteria::ReadPreference(ReadPreference::Primary))
+                .projection(doc! { "_id": 1 })
+                .build();
+            if self
+                .files()
+                .clone_with_type::<Document>()
+                .find_one(None, find_options)
+                .await?
+                .is_none()
+            {
+                Self::create_index(self.files(), doc! { "filename": 1, "uploadDate": 1 }).await?;
+                Self::create_index(self.chunks(), doc! { "files_id": 1, "n": 1 }).await?;
+            }
+            self.inner.created_indexes.store(true, Ordering::SeqCst);
+        }
+
+        Ok(())
+    }
+
+    async fn create_index<T>(coll: &Collection<T>, keys: Document) -> Result<()> {
+        // From the spec: Drivers MUST check whether the indexes already exist before attempting to
+        // create them.
+        let mut indexes = coll.list_indexes(None).await?;
+        'outer: while let Some(index_model) = indexes.try_next().await? {
+            if index_model.keys.len() != keys.len() {
+                continue;
+            }
+            // Indexes should be considered equivalent regardless of numeric value type.
+            // e.g. { "filename": 1, "uploadDate": 1 } is equivalent to
+            // { "filename": 1.0, "uploadDate": 1.0 }
+            let number_matches = |key: &str| index_model.keys.get(key).and_then(get_int) == Some(1);
```

Comment:
I know the indexes we create all currently use 1 for the value, but I think it would be good to avoid future pitfalls by checking against `get_int` of the input document too instead of `Some(1)`.